experimental/whisper/speedup.ipynb

```diff
@@ -0,0 +1,413 @@
+    "    reordered_past = ()\n",
+    "    for layer_past in past:\n",
+    "        reordered_past += (\n",
+    "            tuple(past_state.index_select(0, beam_idx) for past_state in layer_past[:2]) + layer_past[2:],\n",
+    "        )\n",
+    "    return reordered_past\n",
+    "\n",
+    "\n",
+    "WhisperForConditionalGeneration._reorder_cache = fix_reorder_cache"
+   ]
+  },
+  {
+   "attachments": {},
+   "cell_type": "markdown",
+   "metadata": {
+    "collapsed": false
+   },
+   "source": [
+    "### Kernl optimization\n",
+    "\n",
+    "Warmup takes around 12 minutes and is mostly spent by PyTorch 2.0 dynamo module on CPU to capture graph (it is down from 50 min with previous version of Kernl).\n",
+    "We plan to support dynamic shape mode on dynamo, preliminary benchmarks show a 5X faster warmup on Whisper large (basically mostly Triton autotune remains).\n",
+    "\n",
+    "Not that < 2% outputs are different from the original model.\n",
+    "Manual inspection shows that differences are mostly small, many being only one different token in the whole transcription.\n",
+    "\n",
+    "The explanation is that our optimized kernels are mathematically equivalent to the one of the original model but may not perform operations in the same order as PyTorch kernels. As operations on float tensors always lead to rounding, order matters even for commutative operations.\n",
+    "\n",
+    "Moreover, our fused kernels perform accumulations in fp32 which is not possible when you chain PyTorch kernels with fp16 tensors.\n",
+    "When 2 tokens have very similar scores, these rounding differences matters and that's why, in a few cases, outputs are different."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 7,
+   "metadata": {
+    "execution": {
+     "iopub.execute_input": "2023-01-25T19:15:55.378868Z",
+     "iopub.status.busy": "2023-01-25T19:15:55.378673Z",
+     "iopub.status.idle": "2023-01-25T19:49:34.590462Z",
+     "shell.execute_reply": "2023-01-25T19:49:34.589828Z"
+    }
+   },
+   "outputs": [
+    {
+     "name": "stdout",
+     "output_type": "stream",
+     "text": [
+      "time to warmup: 11.58min\n",
+      "timings\n",
+      "[original] average: 1.12s / complete: 48.72min\n",
+      "[optimized] average: 0.46s / complete: 19.87min\n",
+      "# different outputs: 34/2620 (1.30%)\n",
+      "\n",
+      "memory footprint\n",
+      "torch.cuda.memory_allocated: 10.9GB\n",
+      "torch.cuda.memory_reserved: 13.4GB\n",
+      "torch.cuda.max_memory_reserved: 13.9GB\n"
+     ]
+    }
+   ],
+   "source": [
+    "optimize_model(model.model.decoder)\n",
+    "nb_diff = 0\n",
+    "timings_optimized = list()\n",
+    "with torch.inference_mode(), torch.autocast(dtype=torch.float16, cache_enabled=True, device_type=\"cuda\"):\n",
+    "    start = time.time()\n",
+    "    model.generate(inputs_1, min_length=max_len, max_length=max_len, num_beams=num_beams, do_sample=False)\n",
```

Comment:
only 1 ?

Reply:
the warmup generates 51 outputs, enough to warm the PGU